### ISSUE TITLE
bug in weight decay for adam optimizer

### ISSUE THREAD

**Issue by egillax** [mlverse/torch#821]
I found a typo in the adam optimizer loop_fun [here](https://github.com/mlverse/torch/blob/main/R/optim-adam.R#L96)

It is:

```R  
grad$add_(p, alpha = group$weight_decay) 
```

But should be:

```R  
grad$add_(param, alpha = group$weight_decay) 
```
I was comparing the same models in pytorch and torch and found that there were some differences. After a lot of debugging where I tried to overfit one random batch and removed more and more parts of my code I finally identified the weight decay in Adam to be the culprit. With a deep-ish resnet of 8 layers and some weight decay I couldn't overfit one batch, while pytorch did it easily for the same model.

I was going to add a quick pull request to fix it but I'm a bit stuck on how to write a test to catch this..

**Comment by dfalbel** [comment]
Oh wow, I'm so sorry about that and thanks very much @egillax for finding this bug!
I don't think there's a very good way to test it. Maybe having a deterministic example that we can make it return the exact same value as in python. Then we make this value fixed in the test case?
